### DISCUSSION TITLE
How to add a serial port parsing task

### DISCUSSION THREAD

**Discussion by ghost** [pschatzmann/ESP32-A2DP#584] [Q&A] [answered]
I have encountered a problem and after executing your code, I want to add the tasks of serial port receiving and judging, but I do not know how to add them. I have indeed encountered some problems for novices. Hope you can give me some advice

**Comment by pschatzmann** [comment, marked as answer]
I am sorry, but I have no idea what you mean with "serial port receiving and judging".

Finally, if you want to add some serial input or output, this has nothing to do with this library...

**Comment by harryberlin** [comment]
have a look at this. but it's a bit to much for your request.
https://github.com/harryberlin/AVR-IBus.bt

**Comment by pschatzmann** [comment]
Just try to implement and test the functionality w/o A2DP first and when it is working you can add it later on.
The library should not have any impact on the serial Arduino functionality.

**Reply by ghost** [reply]
My English is not very good, I use the translation software to communicate with you, and my programming level is not very high. I added my serial data processing function like this
`char str[100];
void data_received_callback() {
   while (Serial.available() >= 7) {
     unsigned char ubuffer[7];
    
     unsigned char frame_header = Serial.peek();
    
     if (frame_header == 0x55) {
      
       Serial.readBytes(ubuffer, 7);
       if (ubuffer[4] == 0xff && ubuffer[5] == 0xff && ubuffer[6] == 0xff) {
         if(ubuffer[1] == 0x01)
         {
            switch (ubuffer[2])
            {
             case 0x00:
             if(ubuffer[3]==0x01) a2dp_sink.pause();
             if(ubuffer[3]==0x00) a2dp_sink.play();
             break;
              case 0x01: 
                a2dp_sink.previous();
            break;
              case 0x02:
              a2dp_sink.next();
            break;
            }
          }
     } else {
       Serial.read();  
     }
   }
 }
}`
I think this function( a2dp_sink.set_on_data_received(data_received_callback);) is like a timed task so I added some functions I wrote to handle serial port data but this function can only be executed after Bluetooth connection and I don't know if this is good or bad I'm fed up with my programming ability

**Reply by pschatzmann** [reply]
No you would add your logic to the Arduino loop! 
Don't forget to check the result of your Serial.readBytes()

**Reply by ghost** [reply]
I did it. Thank you so much
